src/main/java/org/folio/circulation/support/fetching/PageableFetcher.java

```diff
@@ -0,0 +1,91 @@
+package org.folio.circulation.support.fetching;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.folio.circulation.support.http.client.Offset.noOffset;
+import static org.folio.circulation.support.http.client.PageLimit.limit;
+import static org.folio.circulation.support.results.Result.failed;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.folio.circulation.domain.MultipleRecords;
+import org.folio.circulation.support.ServerErrorFailure;
+import org.folio.circulation.support.http.client.CqlQuery;
+import org.folio.circulation.support.http.client.Offset;
+import org.folio.circulation.support.http.client.PageLimit;
+import org.folio.circulation.support.results.Result;
+import org.slf4j.Logger;
+
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public final class PageableFetcher<T> {
+  private static final Logger log = getLogger(PageableFetcher.class);
+
+  // This limit needed to prevent stack overflow for recursive fetch
+  private static final int DEFAULT_MAX_ALLOWED_RECORDS_LIMIT = 1_000_000;
+  private static final PageLimit DEFAULT_PAGE_SIZE_LIMIT = limit(500);
+
+  private final GetManyRecordsRepository<T> repository;
+  private final PageLimit pageSize;
+  private final int maxAllowedRecordsToFetchLimit;
+
+  public PageableFetcher(GetManyRecordsRepository<T> repository) {
+    this(repository, DEFAULT_PAGE_SIZE_LIMIT, DEFAULT_MAX_ALLOWED_RECORDS_LIMIT);
+  }
+
+  public CompletableFuture<Result<Void>> processPages(CqlQuery query, PageProcessor<T> pageProcessor) {
+    return processPagesRecursively(query, pageProcessor, noOffset());
```

Comment:
I wonder if it might be clearer to have an explicit zero offset, what do you think?

```suggestion
    return processPagesRecursively(query, pageProcessor, zeroOffset());
```

Reply:
Do you mean to rename the existing method or to introduce a new one with offset(0)?

The existing method has following body:

```
public static Offset noOffset() {
    return new Offset(null);
  }
```

Reply:
> Do you mean to rename the existing method or to introduce a new one with offset(0)?

I meant creating a new method for an explicit offset of zero, rather than the omission of an offset at all.

 ```
 public static Offset zeroOffset() {
     return new Offset(0);
   }
 ```

Reply:
Thanks, added.